File: app/src/main/java/com/example/todolistapp/TodoFragment.java
package com.example.todolistapp;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;

import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;


import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.core.content.FileProvider;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Toast;

import java.io.File;
import java.util.UUID;

public class TodoFragment extends Fragment {

    private static final String ARG_TODO_ID = "todo_id";
    private static final int REQUEST_PHOTO = 2;

    private Todo mTodo;
    private EditText mEditTextTitle;
    private Button mButtonDate;
    private CheckBox mCheckBoxIsComplete;

    private ImageButton mPhotoButton;
    private ImageView mPhotoView;
    private File mPhotoFile;

    /*
    Rather than the calling the constructor directly, Activity(s) should call newInstance
    and pass required parameters that the fragment needs to create its arguments.
     */
    public static TodoFragment newInstance(UUID todoId) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_TODO_ID, todoId);

        TodoFragment fragment = new TodoFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);

        /*
         Fragment accessing the intent from the hosting Activity as in the following code snippet
         allows for simple code that works.

        UUID todoId = (UUID) getActivity()
                .getIntent().getSerializableExtra(TodoActivity.EXTRA_TODO_ID);

         The disadvantage: TodoFragment is no longer reusable as it is coupled to Activities whoes
         intent has to contain the todoId.

         Solution: store the todoId in the fragment's arguments bundle.
            See the TodoFragment newInstance(UUID todoId) method.

         Then to create a new fragment, the TodoActivity should call TodoFragment.newInstance(UUID)
         and pass in the UUID it retrieves from its extra argument.

        */

        UUID todoId = (UUID) getArguments().getSerializable(ARG_TODO_ID);
        mTodo = TodoModel.get(getActivity()).getTodo(todoId);
        mPhotoFile = TodoModel.get(getActivity()).getPhotoFile(mTodo);

    }

    @Override
    public void onPause() {
        super.onPause();
        TodoModel.get(getActivity()).updateTodo(mTodo);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_todo, container, false);

        mEditTextTitle = (EditText) view.findViewById(R.id.todo_title);
        mEditTextTitle.setText(mTodo.getTitle());
        mEditTextTitle.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // This line is intentionally left blank
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                mTodo.setTitle(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {
                // This line is intentionally left blank
            }
        });

        mButtonDate = (Button) view.findViewById(R.id.todo_date);
        mButtonDate.setText(mTodo.getDate().toString());
        mButtonDate.setEnabled(false);

        mCheckBoxIsComplete = (CheckBox) view.findViewById(R.id.todo_complete);
        mCheckBoxIsComplete.setOnCheckedChangeListener(new OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                Log.d("DEBUG **** TodoFragment", "called onCheckedChanged");
                mTodo.setComplete(isChecked==true ? 1 : 0);
            }
        });

        mPhotoButton = (ImageButton) view.findViewById(R.id.todo_camera);
        mPhotoView = (ImageView) view.findViewById(R.id.todo_photo);

        final Intent captureImage = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);



        PackageManager packageManager = getActivity().getPackageManager();
        boolean canTakePhoto = mPhotoFile != null && packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY);
        mPhotoButton.setEnabled(canTakePhoto);

        ActivityResultLauncher<Uri> todoTakePicture = registerForActivityResult(
                new ActivityResultContracts.TakePicture(),
                result -> {
                    if (result) {
                        Uri uri = FileProvider.getUriForFile(TodoFragment.this.getActivity(),
                                "com.example.todolistapp.fileprovider",
                                mPhotoFile);

                        TodoFragment.this.getActivity().revokeUriPermission(uri, Intent.FLAG_GRANT_WRITE_URI_PERMISSION);

                        TodoFragment.this.updatePhotoView();
                    } else {
                        return;
                    }

                });



        mPhotoButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if(checkCameraPermission()) {
                    Uri uri = FileProvider.getUriForFile(getActivity(),
                            "com.example.todolistapp.fileprovider",
                            mPhotoFile);

                    todoTakePicture.launch(uri);
                }
                else {
                    requestPermissionLauncher.launch(Manifest.permission.CAMERA);
                }
            }
        });

        mPhotoView = (ImageView) view.findViewById(R.id.todo_photo);
        updatePhotoView();

        return view;

    }

    private boolean checkCameraPermission() {
        if (ContextCompat.checkSelfPermission(
                this.getActivity().getApplicationContext(), Manifest.permission.CAMERA) ==
                PackageManager.PERMISSION_GRANTED)
        {
            return true;
        }
        else {
            return false;
        }
    }

    private void updatePhotoView(){
        if (mPhotoFile == null || !mPhotoFile.exists()) {
            mPhotoView.setImageDrawable(null);
        } else {
            Bitmap bitmap = PictureUtils.getScaledBitmap(mPhotoFile.getPath(), getActivity());
            mPhotoView.setImageBitmap(bitmap);
        }
    }

    // Register the permissions callback, which handles the user's response to the
    // system permissions dialog. Save the return value, an instance of
    // ActivityResultLauncher, as an instance variable.
    private ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {

                } else {
                    Toast.makeText(
                            TodoFragment.this.getActivity(),
                            "Permission to use camera has not been granted",
                            Toast.LENGTH_SHORT)
                            .show();
                }
            });

}

File: app/src/main/java/com/example/todolistapp/TodoFragment.kt
package com.example.todolistapp

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import java.io.File
import java.util.*

class TodoFragment : Fragment() {
    private var mTodo: Todo? = null
    private lateinit var mEditTextTitle: EditText
    private lateinit var mButtonDate: Button
    private lateinit var mCheckBoxIsComplete: CheckBox
    private lateinit var mPhotoButton: ImageButton
    private lateinit var mPhotoView: ImageView
    private var mPhotoFile: File? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        /*
         Fragment accessing the intent from the hosting Activity as in the following code snippet
         allows for simple code that works.

        UUID todoId = (UUID) getActivity()
                .getIntent().getSerializableExtra(TodoActivity.EXTRA_TODO_ID);

         The disadvantage: TodoFragment is no longer reusable as it is coupled to Activities whoes
         intent has to contain the todoId.

         Solution: store the todoId in the fragment's arguments bundle.
            See the TodoFragment newInstance(UUID todoId) method.

         Then to create a new fragment, the TodoActivity should call TodoFragment.newInstance(UUID)
         and pass in the UUID it retrieves from its extra argument.

        */
        val todoId = requireArguments().getSerializable(ARG_TODO_ID) as UUID?
        mTodo = TodoModel.get(activity)!!.getTodo(todoId)
        mPhotoFile = TodoModel.get(activity)!!.getPhotoFile(mTodo, context)
    }

    override fun onPause() {
        super.onPause()
        TodoModel.get(activity)!!.updateTodo(mTodo)
    }

    @SuppressLint("UseRequireInsteadOfGet")
    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_todo, container, false)
        mEditTextTitle = view.findViewById<View>(R.id.todo_title) as EditText
        mEditTextTitle.setText(mTodo!!.title)
        mEditTextTitle!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // This line is intentionally left blank
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                mTodo!!.title = s.toString()
            }

            override fun afterTextChanged(s: Editable) {
                // This line is intentionally left blank
            }
        })
        mButtonDate = view.findViewById(R.id.todo_date)
        mButtonDate.setText(mTodo!!.date.toString())
        mButtonDate.isEnabled = false
        mCheckBoxIsComplete = view.findViewById(R.id.todo_complete)
        mCheckBoxIsComplete.setOnCheckedChangeListener { buttonView, isChecked ->
            Log.d("DEBUG **** TodoFragment", "called onCheckedChanged")
            mTodo!!.isComplete = if (isChecked == true) true else false
        }
        mPhotoButton = view.findViewById<View>(R.id.todo_camera) as ImageButton
        mPhotoView = view.findViewById<View>(R.id.todo_photo) as ImageView
        val captureImage = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        val packageManager = requireActivity().packageManager
        val canTakePhoto = mPhotoFile != null && packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)
        mPhotoButton.isEnabled = canTakePhoto
        val todoTakePicture = registerForActivityResult<Uri, Boolean>(
                ActivityResultContracts.TakePicture()
        ) { result: Boolean ->
            if (result) {
                val uri = FileProvider.getUriForFile(this@TodoFragment.activity!!,
                        "com.example.todolistapp.fileprovider",
                        mPhotoFile!!)
                this@TodoFragment.activity!!.revokeUriPermission(uri, Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
                updatePhotoView()
            } else {
                return@registerForActivityResult
            }
        }
        mPhotoButton.setOnClickListener {
            if (checkCameraPermission()) {
                val uri = FileProvider.getUriForFile(activity!!,
                        "com.example.todolistapp.fileprovider",
                        mPhotoFile!!)
                todoTakePicture.launch(uri)
            } else {
                requestPermissionLauncher.launch(Manifest.permission.CAMERA)
            }
        }
        mPhotoView = view.findViewById<View>(R.id.todo_photo) as ImageView
        updatePhotoView()
        return view
    }

    private fun checkCameraPermission(): Boolean {
        return if (ContextCompat.checkSelfPermission(
                        this.requireActivity().applicationContext, Manifest.permission.CAMERA) ==
                PackageManager.PERMISSION_GRANTED) {
            true
        } else {
            false
        }
    }

    private fun updatePhotoView() {
        if (mPhotoFile == null || !mPhotoFile!!.exists()) {
            mPhotoView.setImageDrawable(null)
        } else {
            val bitmap = PictureUtils.getScaledBitmap(mPhotoFile!!.path, activity)
            mPhotoView.setImageBitmap(bitmap)
        }
    }

    // Register the permissions callback, which handles the user's response to the
    // system permissions dialog. Save the return value, an instance of
    // ActivityResultLauncher, as an instance variable.
    private val requestPermissionLauncher = registerForActivityResult<String, Boolean>(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
        if (isGranted) {
        } else {
            Toast.makeText(
                    this@TodoFragment.activity,
                    "Permission to use camera has not been granted",
                    Toast.LENGTH_SHORT)
                    .show()
        }
    }

    companion object {
        private const val ARG_TODO_ID = "todo_id"
        private const val REQUEST_PHOTO = 2

        /*
    Rather than the calling the constructor directly, Activity(s) should call newInstance
    and pass required parameters that the fragment needs to create its arguments.
     */
        fun newInstance(todoId: UUID?): TodoFragment {
            val args = Bundle()
            args.putSerializable(ARG_TODO_ID, todoId)
            val fragment = TodoFragment()
            fragment.arguments = args
            return fragment
        }
    }
}